### PR TITLE
[Merged by Bors] - feat: port smul positivity extension

### PR DESCRIPTION
**Pull request by dwrensha**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Order/SMul.lean [hunk 1]
```diff
@@ -364,54 +364,59 @@ private theorem smul_ne_zero_of_ne_zero_of_pos [Preorder M] (ha : a ≠ 0) (hb :
...
-end Tactic
+open Lean.Meta Qq
+
+/-- Positivity extension for HSMul, i.e. (_ • _).  -/
+@[positivity HSMul.hSMul _ _]
+def evalHSMul : PositivityExt where eval {_u α} zα pα (e : Q($α)) := do
+  let .app (.app (.app (.app (.app (.app
+        (.const ``HSMul.hSMul [u1, _, _]) (M : Q(Type u1))) _) _) _)
+          (a : Q($M))) (b : Q($α)) ← whnfR e | throwError "failed to match hSMul"
+  let .some zM ← trySynthInstance (q(Zero $M))
```
Comment (dwrensha) [inline Mathlib/Algebra/Order/SMul.lean:375]: If I change this to
```lean
  let zM ← synthInstanceQ (q(Zero $M))
  ```
  then I get
```
Messages below (1)
378:2: 
(kernel) declaration has free variables 'Mathlib.Meta.Positivity.evalHSMul.match_1'
```

File: Mathlib/Algebra/Order/SMul.lean [hunk 2]
```diff
@@ -364,54 +364,59 @@ private theorem smul_ne_zero_of_ne_zero_of_pos [Preorder M] (ha : a ≠ 0) (hb :
...
+    | throwError "failed to synthesize `Zero $M`"
+  let .some pM ← trySynthInstance (q(PartialOrder $M))
+    | throwError "failed to synthesize `PartialOrder $M`"
+  match ← core zM pM a, ← core zα pα b with
+  | .positive pa, .positive pb =>
+      pure (.positive (← mkAppM ``smul_pos #[pa, pb]))
```
Comment (dwrensha) [inline Mathlib/Algebra/Order/SMul.lean:381]: If I try to use `q()` here, I get a weird error:
```
381:23: 
unquoteExpr: pM : Lean.Expr
```

File: Mathlib/Algebra/Order/SMul.lean (hunk 1 above)
Comment (dwrensha) [reply Mathlib/Algebra/Order/SMul.lean]: fixed via type ascription

File: test/positivity.lean [hunk 3]
```diff
@@ -144,12 +144,22 @@ example {a : ℚ} (ha : 0 < a) : 0 < |a| := by positivity
...
+example {a : ℤ} {b : ℚ} (ha : a ≠ 0) (hb : 0 < b) : a • b ≠ 0 := by positivity
+example {a : ℤ} {b : ℚ} (ha : a ≠ 0) (hb : b ≠ 0) : a • b ≠ 0 := by positivity
+
+example {a : ℤ} {b : ℚ} (ha : 0 < a) (hb : 0 < b) : 0 < SMul.smul a b := by positivity
+example {a : ℤ} {b : ℚ} (ha : 0 < a) (hb : 0 ≤ b) : 0 ≤ SMul.smul a b := by positivity
+example {a : ℤ} {b : ℚ} (ha : 0 ≤ a) (hb : 0 < b) : 0 ≤ SMul.smul a b := by positivity
+example {a : ℤ} {b : ℚ} (ha : 0 ≤ a) (hb : 0 ≤ b) : 0 ≤ SMul.smul a b := by positivity
+example {a : ℤ} {b : ℚ} (ha : 0 < a) (hb : b ≠ 0) : SMul.smul a b ≠ 0 := by positivity
+example {a : ℤ} {b : ℚ} (ha : a ≠ 0) (hb : 0 < b) : SMul.smul a b ≠ 0 := by positivity
+example {a : ℤ} {b : ℚ} (ha : a ≠ 0) (hb : b ≠ 0) : SMul.smul a b ≠ 0 := by positivity
```
Comment (eric-wieser) [inline test/positivity.lean:162]: Can you add a test to check that the universe polymorphism is correct in the tactic? I'm thing a super general example stated with `OrderedSMul`, with types in separate universes.

File: Mathlib/Algebra/Order/SMul.lean [hunk 4]
```diff
@@ -364,54 +364,55 @@ private theorem smul_ne_zero_of_ne_zero_of_pos [Preorder M] (ha : a ≠ 0) (hb :
...
+      pure (.nonzero (← mkAppM ``smul_ne_zero #[pa, pb]))
+  | _, _ => pure .none
+
+/-- Positivity extension for SMul. -/
+@[positivity SMul.smul _ _]
+def evalSMul : PositivityExt where eval {_u α} zα pα (e : Q($α)) := do
+  let .app (.app (.app (.app (.app
+        (.const ``SMul.smul [u1, _]) (M : Q(Type u1))) _) _)
+          (a : Q($M))) (b : Q($α)) ← whnfR e | throwError "failed to match smul"
+  -- TODO: deduplicate this logic, as it's shared with the HSMul extension, above.
```
Comment (eric-wieser) [inline Mathlib/Algebra/Order/SMul.lean:400]: This shouldn't ever be needed: `SMul.smul` is an implementation detail.

Comment (dwrensha) [reply Mathlib/Algebra/Order/SMul.lean:400]: `evalAdd` is covers both `HAdd.hAdd` and `Add.add`:

https://github.com/leanprover-community/mathlib4/blob/00cb9455117f9b1046c7fa0d15bf5b8ad0654c1c/Mathlib/Tactic/Positivity/Basic.lean#L156

and `evalMul` covers both `HMul.mMul` and `Mul.mul`:
https://github.com/leanprover-community/mathlib4/blob/00cb9455117f9b1046c7fa0d15bf5b8ad0654c1c/Mathlib/Tactic/Positivity/Basic.lean#L197


Is there anything different about `SMul` that says we *don't* need to cover both cases here? (I would put `HSMul.hSMul` and `SMul.smul` in a single extension if I could figure out how.)

Comment (eric-wieser) [reply Mathlib/Algebra/Order/SMul.lean:400]: I think all of these are mistakes, and could also be removed (in a separate PR)

File: test/positivity.lean (hunk 3 above)
Comment (dwrensha) [reply test/positivity.lean:167]: Done in https://github.com/leanprover-community/mathlib4/pull/8067/commits/74dee76132621af52732b65fa64f5cc1b78af04e.

File: Mathlib/Algebra/Order/SMul.lean (hunk 4 above)
Comment (dwrensha) [reply Mathlib/Algebra/Order/SMul.lean]: Removed the SMul extension in https://github.com/leanprover-community/mathlib4/pull/8067/commits/7876e072c6441fe1bfc33c0a309a13d8d95752b4

**Comment by eric-wieser** [comment]
bors merge

Thanks! On reflection I think using Qq here would be a bad idea, as it requires us to hard-code all the instances we think we need, meaning we risk forgetting to mirror generalizations made to the lemmas in future.

**Pull request closed without merging**